cluster-autoscaler/utils/drain/drain.go

```diff
@@ -250,10 +253,19 @@ func isPodTerminal(pod *apiv1.Pod) bool {
 	return pod.Status.Phase == apiv1.PodFailed
 }
 
-// HasLocalStorage returns true if pod has any local storage.
-func HasLocalStorage(pod *apiv1.Pod) bool {
+// HasBlockingLocalStorage returns true if pod has any local storage
+// without pod annotation `<SafeToEvictLocalVolumeKey>: <volume-name-1>,<volume-name-2>...`
+func HasBlockingLocalStorage(pod *apiv1.Pod) bool {
+	annotationVal := pod.GetAnnotations()[SafeToEvictLocalVolumeKey]
+	isNonBlocking := map[string]bool{}
+	if annotationVal != "" {
+		vols := strings.Split(annotationVal, ",")
+		for _, v := range vols {
+			isNonBlocking[v] = true
```

Comment:
I am creating a map here which seems a little wasteful. Any better suggestions are welcome :pray:

Reply:
:memo: thank you for the explanation. 

Say we have a lot of pods which need to be drained from a node marked for scale down. Would it be fair to say, since we drain only one node at a time, the checking we do for every pod is still not as performance heavy as API calls and binpacking complexity.

Reply:
Yes exactly, it should be negligible in practice. Also, k8s imposes limits on the maximum number of pods on a node (default is 110 I think? At least in GKE it's somewhere in that ballpark), and the actual number is in practice much lower than the limit, due to many reasons.

Also FYI, we have parallel drain in the works, it'll probably be turned on by default in the next CA release.